import sys
from PyQt5.QtWidgets import *
from PyQt5.QtGui import *
from PyQt5 import uic
from collections import Counter
from_class = uic.loadUiType("calculator.ui")[0]

class WindowClass(QMainWindow, from_class) :
    def __init__(self):
        self.stack = []
        self.operators = "+-*/(."
        self.numbers = "0123456789"
        ac_state = False
        super().__init__()
        self.setupUi(self)
        self.setWindowTitle("Calculator")
        button_functions = {
            'pushButton_0': self.button_0_clicked,
            'pushButton_1': self.button_1_clicked,
            'pushButton_2': self.button_2_clicked,
            'pushButton_3': self.button_3_clicked,
            'pushButton_4': self.button_4_clicked,
            'pushButton_5': self.button_5_clicked,
            'pushButton_6': self.button_6_clicked,
            'pushButton_7': self.button_7_clicked,
            'pushButton_8': self.button_8_clicked,
            'pushButton_9': self.button_9_clicked,
            'pushButton_point' : self.point,
            'pushButton_back' : self.backspace,
            'pushButton_allclear' : self.allclear,
            'pushButton_plus' : self.plus,
            'pushButton_minus' : self.minus,
            'pushButton_multiply' : self.multiply,
            'pushButton_divide' : self.divide,
            'pushButton_result' : self.calc_result,
            'pushButton_startbracket' : self.start_bracket,
            'pushButton_endbracket' : self.end_bracket,
            'pushButton_pi' : self.pi,
            'pushButton_mod' : self.mod,
            'pushButton_root' : self.root,
            'pushButton_exp' : self.exp
            # ÎÇòÎ®∏ÏßÄ Î≤ÑÌäºÎì§Ïóê ÎåÄÌï¥ÏÑúÎèÑ Îß§Ìïë Ï∂îÍ∞Ä
        }

        self.mode = {
            'pushButton_point' : '.',
            'pushButton_plus' : '+',
            'pushButton_minus' : '-',
            'pushButton_multiply' : '*',
            'pushButton_divide' : '/'
        }

        

        for button_name, button_function in button_functions.items():
            # if button_name in self.mode: 
            #     self.state = button_name
            #     check = 1
            button = getattr(self, button_name)
            if ac_state == True and (button_name == 'pushButton_allclear'):
                ac_state = False
                self.Window.clear()


            if button_name == 'pushButton_allclear':
                ac_state = True
                button.clicked.connect(button_function)
            else:
                ac_state = False
                button.clicked.connect(button_function)
            

                
    def split_expression(self,expression):
        operators = ['+','-','*','/','(',')','^','‚àö']
        result = []
        i = 0

        while i < len(expression):
            if expression[i] == ' ':
                i+=1
                continue
            if expression[i].isdigit():
                j = i
                while j < len(expression) and (expression[j].isdigit() or expression[j] == '.'):
                    j += 1
                result.append(expression[i:j])
                i = j
            elif expression[i] in operators:
                result.append(expression[i])
                i += 1
            elif expression[i] == 'ùùø':
                result[-1] = str(float(result[-1])*3.141592)
                i += 1
            elif expression[i] == 'm':
                result.append('%')
                i += 3
            else:
                raise ValueError("Invalid character in expression: " + expression[i])
        return result 
    

    def isNumeric(self,num):
        try:
            float(num)
            return True
        except ValueError:
            return False
        
    def calc_result(self):
        # document = self.Window.document()
        # # ÎßàÏßÄÎßâ Ï§ÑÏùò ÌÖçÏä§Ìä∏ Í∞ÄÏ†∏Ïò§Í∏∞
        # last_block = document.findBlockByLineNumber(document.blockCount())
        # last_line_text = last_block.text()
        last_line_text = self.Window.text()
        tokens = self.split_expression(last_line_text)
        c = Counter(tokens)
        if c['('] != c[')']:
             self.Window_2.setText('ÏóêÎü¨ : Í¥ÑÌò∏ Í∞ØÏàò ÌôïÏù∏.')
             raise Exception("Í¥ÑÌò∏Ïùò Í∞ØÏàòÍ∞Ä ÎßûÏßÄ ÏïäÏäµÎãàÎã§. ÌôïÏù∏Ìï¥Ï£ºÏÑ∏Ïöî.")

        lst = []        # Îπà Î¶¨Ïä§Ìä∏ ÏÉùÏÑ±
        stack = []      # Ïä§ÌÉù ÏÉùÏÑ±

        prior = {'‚àö': 4,'^':4, '%':3, '*':3, '/':3, '+':2, '-':2, '(':1}     # Ïö∞ÏÑ†ÏàúÏúÑ ÏÑ§Ï†ï
        for n in range(len(tokens)):    # ÌÜ†ÌÅ∞Ïùò Í∏∏Ïù¥ÎßåÌÅº Î∞òÎ≥µÌïòÏó¨
            prior = {'‚àö': 4,'^':4, '%':3, '*':3, '/':3, '+':2, '-':2, '(':1} 
            if self.isNumeric(tokens[n]) : # ÎßåÏïΩ Ïà´ÏûêÏù¥Î©¥ Î∞îÎ°ú lstÏóê Ï∂îÍ∞Ä
                lst.append(tokens[n])
            elif tokens[n] == '(':  # (Ïù¥Î©¥ Î∞îÎ°ú stackÏóê Ï∂îÍ∞Ä
                    stack.append(tokens[n])
            elif tokens[n] == ')':  # )Í∞Ä ÎÇòÏò§Î©¥ stackÏóêÏÑú (Í∞Ä ÎÇòÏò¨ÎïåÍπåÏßÄ popÏ≤òÎ¶¨ Î∞è lstÏóê Ï∂îÍ∞Ä. 
                while stack[-1] != '(':
                    lst.append(stack.pop())
                stack.pop() # (Í∞Ä ÎÇòÌÉÄÎÇòÎ©¥ popÏ≤òÎ¶¨
            else:   # Í∑∏Ïô∏Ïóê Í≤ΩÏö∞ tokens[n]Ïù¥ stack[-1]Ïùò Ïö∞ÏÑ†ÏàúÏúÑÏôÄ Í∞ôÍ±∞ÎÇò Î≥¥Îã§ ÏûëÏúºÎ©¥ tokens[n]Ïùò Ïö∞ÏÑ†ÏàúÏúÑÍ∞Ä Îçî Ïª§ÏßàÎïåÍπåÏßÄ pop
                while stack and prior[tokens[n]] <= prior[stack[-1]]:
                    lst.append(stack.pop()) # popÌïúÍ≤ÉÎì§ÏùÄ lstÏóê Ï∂îÍ∞Ä ÏãúÏºúÏ§å   
                stack.append(tokens[n]) # ÏúÑÏùò Ï°∞Í±¥Ïù¥ ÏôÑÎ£å ÎêòÎ©¥ stackÏóê Ï∂îÍ∞Ä

        while len(stack) != 0:  # stackÏóê ÎÇ®ÏïÑÏûàÎäî Ïó∞ÏÇ∞ÏûêÎì§ lstÏóê Ï∂îÍ∞Ä
            lst.append(stack.pop())


        # op_check = ['+','/','*','-']    # Ïó∞ÏÇ∞Ïûê Ï≤¥ÌÅ¨Î•º ÏúÑÌï¥ ÎØ∏Î¶¨ ÏÉùÏÑ±
        stack =[]   # ÌîºÏó∞ÏÇ∞Ïûê Î∞îÎ°ú Ï∂îÍ∞ÄÌï† Î¶¨Ïä§Ìä∏ ÏÉùÏÑ±
        a1=0   # stack[-1]ÏùÑ ÏúÑÌïú Î≥ÄÏàò ÏÉùÏÑ±
        a2=0  # stack[-2]ÏùÑ ÏúÑÌïú Î≥ÄÏàò ÏÉùÏÑ±
        for l in range(len(lst)):   # ÌõÑÏúÑÌëúÍ∏∞Î≤ïÏúºÎ°ú Ï†ÄÏû•Îêò ÏûàÎäî Î¶¨Ïä§Ìä∏Ïùò ÏàòÎßåÌÅº Î∞òÎ≥µ   
            if self.isNumeric(lst[l]):    # ÎßåÏïΩ ÌîºÏó∞ÏÇ∞ÏûêÏù¥Î©¥ Î∞îÎ°ú stackÏóê Ï∂îÍ∞Ä
                stack.append(float(lst[l]))
            elif lst[l] == '^':
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1**a2)
            elif lst[l] == '‚àö':
                a1 = stack.pop()
                stack.append(pow(a1,1/2))
            
            elif lst[l] == '%':
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1 % a2)
            elif lst[l] == '+': # + Ïù¥Î©¥ stackÏóêÏÑú 2Í∞ú ÌîºÏó∞ÏÇ∞ÏûêÎ•º popÌïòÏó¨ Í≤åÏÇ∞Ìï¥Ï§ÄÎí§ Îã§Ïãú stackÏóê Ï∂îÍ∞Ä
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1 + a2)
            elif lst[l] == '-': # - Ïù¥Î©¥ stackÏóêÏÑú 2Í∞ú ÌîºÏó∞ÏÇ∞ÏûêÎ•º popÌïòÏó¨ Í≤åÏÇ∞Ìï¥Ï§ÄÎí§ Îã§Ïãú stackÏóê Ï∂îÍ∞Ä
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1 - a2) 
            elif lst[l] == '*': # * Ïù¥Î©¥ stackÏóêÏÑú 2Í∞ú ÌîºÏó∞ÏÇ∞ÏûêÎ•º popÌïòÏó¨ Í≤åÏÇ∞Ìï¥Ï§ÄÎí§ Îã§Ïãú stackÏóê Ï∂îÍ∞Ä
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1 * a2)
            elif lst[l] == '/': # / Ïù¥Î©¥ stackÏóêÏÑú 2Í∞ú ÌîºÏó∞ÏÇ∞ÏûêÎ•º popÌïòÏó¨ Í≤åÏÇ∞Ìï¥Ï§ÄÎí§ Îã§Ïãú stackÏóê Ï∂îÍ∞Ä
                a2 = stack.pop()
                a1 = stack.pop()
                stack.append(a1 / a2)

        self.Window_2.setText(str(stack[0]))


    def exp(self):
        self.current_text = self.Window.text()
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + '^'
                self.Window.setText(new_text)            

    def root(self):
        self.current_text = self.Window.text()
        # if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
        new_text = self.current_text + ' ‚àö'
        self.Window.setText(new_text)          

    def mod(self):
        self.current_text = self.Window.text()
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + ' mod '
                self.Window.setText(new_text)         

    def pi(self):
        self.current_text = self.Window.text()
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + 'ùùø'
                self.Window.setText(new_text)


    def start_bracket(self):
        self.current_text = self.Window.text()
        # new_text = self.current_text + '('
        # self.noRepeatSym_input(self,new_text)
        if self.current_text[-1] in self.operators:
            new_text = self.current_text + '('
            self.Window.setText(new_text)
   

    def end_bracket(self):
        self.current_text = self.Window.text()
        if self.current_text[-1].isdigit():
                new_text = self.current_text + ')'
                self.Window.setText(new_text)
        
        # self.noRepeatSym_input(self,new_text)
        
        # self.Window.setText(new_text)            

    def point(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        
        if (self.current_text[-1].isdigit() or self.current_text[-1] == ')') :
                new_text = self.current_text + '.'
                self.Window.setText(new_text)
        # if self.current_text[-1].isdigit():
        #     new_text = self.current_text + '.'
        #     self.Window.setText(new_text)

    def plus(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '+'
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + '+'
                self.Window.setText(new_text)
        # self.noRepeatSym_input(new_text)
        # self.Window.setText(new_text)
    
    def minus(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '-'
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + '-'
                self.Window.setText(new_text)
        # self.noRepeatSym_input(new_text)
        # self.Window.setText(new_text)

    def multiply(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '*'
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + '*'
                self.Window.setText(new_text)
        # self.noRepeatSym_input(new_text)
        # self.Window.setText(new_text)

    def divide(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '/'
        if self.current_text[-1].isdigit() or self.current_text[-1] == ')':
                new_text = self.current_text + '/'
                self.Window.setText(new_text)
        # self.noRepeatSym_input(new_text)
        # self.Window.setText(new_text)                

    def allclear(self):
        self.Window.setText('')
        self.Window_2.setText('')

    def backspace(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text[:-1]
        # self.Window.setText(new_text)
        self.Window.setText(new_text)

    def button_0_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '0'
        self.Window.setText(new_text)

    def button_1_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '1'
        self.Window.setText(new_text)   

    def button_2_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '2'
        self.Window.setText(new_text)   

    def button_3_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '3'
        self.Window.setText(new_text)   

    def button_4_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '4'
        self.Window.setText(new_text)   

    def button_5_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '5'
        self.Window.setText(new_text)   

    def button_6_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '6'
        self.Window.setText(new_text)   

    def button_7_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '7'
        self.Window.setText(new_text)   

    def button_8_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '8'
        self.Window.setText(new_text)   

    def button_9_clicked(self):
        # self.current_text = self.Window.toPlainText()
        self.current_text = self.Window.text()
        new_text = self.current_text + '9'
        self.Window.setText(new_text)   

    

if __name__ == "__main__":

    app = QApplication(sys.argv)
    myWindows = WindowClass()
    myWindows.show()
    sys.exit(app.exec_())
